### DISCUSSION TITLE
Automatic navigation of links, with precautions. Demo scripts.

### DISCUSSION THREAD

**Discussion by JJSlote** [freeplane/freeplane#832] [Show and tell]
Hi -- Occasional Freeplane user from way back. I have quite a few HTML documents saved and managed in Scrivener, and markdown documents created in Obsidian. Planning a Freeplane dashboard for these that will allow for descriptions, cross references, highlighting and meta-data. But like some other users, I've been aiming for a cautious way to navigate to node links automatically upon node selection.

These two scripts, in Groovy and AutoHotkey, actually work pretty well to navigate the links, but some of the mechanisms deployed are quite inelegant, as you'll see:

```
// @ExecutionModes({ON_SINGLE_NODE})
def myFile = new File('h:/FreeplaneNodeLink.txt')
myFile.write("FPLink " + node.link.text)
```

```
TimerMMLinks:
PrevMMLink := ""
ffcmd := chr(34) . "C:\Program Files\Mozilla Firefox\firefox.exe" . chr(34)
SetTimer, MMLinks, 1000, 0
Return


MMLinks:
SetTitleMatchMode, 2
ActiveID := Winactive("Freeplane")
if ActiveID 
	{
send, ^{F11}
sleep, 100
fileread, scrap, h:\FreeplaneNodeLink.txt
if instr(scrap, "FPLink ") {
scrap := strreplace(scrap,"FPLink ")
if ((strlen(trim(scrap)) and (scrap <> PrevMMLink)) and (scrap <> "null"))
 {
 run, %ffcmd% %scrap%
PrevMMLink := scrap
 } 
}
}
Return
```

In this scripting tandem:
- Drive H is a RAM drive, created via IMDisk.
- The Groovy script simply writes the selected node's link out to a file on the drive. It's run, for now, on a hotkey, Ctrl-F11.
- The AutoHotkey script times the press of the hotkey to trigger the broadcast of the link. It then reads the file generated by the Groovy script, checks for a fresh and valid link, and adds it as a navigation parameter for Firefox.
- For display, Firefox works well:
     - It knows it's already running, even when called explicitly.
     - It updates the file display in the existing tab rather than creating a new one. 
- I'm using AutoHotkey primarily for familiarity, but also because timed loops in Groovy scripts tend to lock Freeplane up, whereas those in Autohotkey operate in the background for virtual multithreading.

As for the improvements I'm seeking:

1. Surely there's a better way to trigger the communication of the link. Automating the press of a hotkey to broadcast the link info is just my beginner's approach. I've seen intimations of a node selection listener, but couldn't find any such option in the API readout. Automated pressing of the hotkey on a timer causes an annoying hourglass flash each second the Freeplane window is active.

2. I'd like a better way for my Freeplane script to communicate the selected node's link info to the outside.

- A ram drive is an easy mechanism for cross-program communication, but it shouldn't be necessary. I happen to be running one anyway, so for inter-program communication it works well for me, but most users would not care to set one up. Obviously we don't want to update a file on a physical drive anywhere near that frequently. On the other hand, I do have a mechanism in AutoHotkey to monitor selected folders and files for changes, so if a file could be updated on a new node selection, it would obviate the need for timers in the scripts.

- A window caption is an elegant, if unconventional, way of sharing data between scripts. WinExist() in AHK can find a window on a partial string match, enabling the script to parse the full caption. Freeplane does have a menu option to change the window caption (workspace title) but it appears limited to fixed text. Freeplane also uses a couple of other non-modal windows whose titles can be identified and put to use by outside applications. (_Search and Replace_ and _Manage Reminders_ illustrate the concept.) If the status bar, for example, could be rendered in its own window, with contents as caption, my AHK script could easily interrogate it. On the other hand, I'd still have to run a timer in AutoHotkey to respond to changes in a Freeplane window caption.
 
- An environment variable is another possibility for sharing data between scripts. Yet to be explored, as the ram drive works well for my purposes. (Certainly the clipboard is out; it's used actively and we don't want it overwritten by an iterative, automated process.)

These scripts are a working start, and now I'd appreciate any insights from Freeplane's community of experts on ways to enhance and streamline them!

Cheers  - Jerome

**Comment by sifran-github** [comment]
i dont really understand what you are trying to do...? can you please describe better what is your goal here?

for example, i dont really understand why you are writing the link into a file and then opening it... instead of... just clicking the link? i read it quickly, maybe i missed the point, but its not clear.

**Reply by JJSlote** [reply]
> i dont really understand why you are writing the link into a file and then opening it... instead of... just clicking the link? 

The map is primarily a collection of documents, and I sought a means whereby merely selecting the node would be sufficient to preview the corresponding document, just as in Obsidian or Scrivener or File Explorer. Indeed, Freeplane gives us the option to keep the Note Panel open or not; similarly, this is a toggled preview pane for linked files. I know it's not a new idea; another user in this forum [suggested](https://github.com/freeplane/freeplane/discussions/429) a link preview capability on a node hover. 

Previewing under script gives me a lot more control as well; for security I can preview by an explicit call to Firefox, and with the link spec captured I can use another hotkey to open the document for editing.

Anyway, the script serves the purpose for which it was created. I'm now seeking to add some sophistication to it, and other scripters might see the benefit as well.

Rgds - Jerome

**Comment by euu2021** [comment]
Interesting!

My first thought is to try calling firefox directly from the FP script, and passing the link as a parameter. But, I don't know if it's possible. Doing a quick search, I found [this discussion](https://sourceforge.net/p/freeplane/discussion/758437/thread/00fc7331/), where Dimitry said:

> Currently Freeplane does not support application calls with parameters explicitly. But I think it is possible to write a script which transforms content of node link or some node attribute to such a call and assign a short cut to this script so that it can be started easy.

So, I think it is a promising path to investigate.

On a side note: for previewing, I would also try, [QuickLook](https://github.com/QL-Win/QuickLook), which also allows this kind of integration by parameter passing. I would try it to see if the operation gets overall lighter.

**Reply by euu2021** [reply]
I read it again, and it's simpler than I thought. In the discussion mentioned above, Volker already gave the answer. So, this worked for me:

```
"\"C:/Program Files/Google/Chrome/Application/chrome.exe\" ${node.link.text}".execute()
```

**Reply by euu2021** [reply]
Now, it seems that you also want to make the preview appear just by selecting the node (without clicking any icon or using hotkeys to activate the script), right? If so, I think the Node Change listeners are the way to go, coupled with the #688 addon, by macmarrum.

I still did not master the art of building Node Listeners, so I can't help much about that.

Also, there are other ways of doing that, as mentioned in #544.

**Reply by euu2021** [reply]
At first, I thought that Node Listeners would solve your problem, because, when I used [this](https://sourceforge.net/p/freeplane/discussion/758437/thread/e9200db6c3/?limit=25#01b5) script by macmarrum, I saw that it reacts to simple node selection. But, testing now, I only got reaction on first selection after creation (for example, after pasting the node). I modified the script and tested like this:

```

import org.freeplane.api.NodeChangeListener
import org.freeplane.api.NodeChanged
import org.freeplane.plugin.script.FreeplaneScriptBaseClass.ConfigProperties

class MyNodeChangeListener implements NodeChangeListener {
    public void nodeChanged(NodeChanged  event) {
        if (!event.node.root && event.node.link.text == null)
            event.node.setMinimized(false)

        else
                "\"C:/Program Files/Google/Chrome/Application/chrome.exe\" ${event.node.link.text}".execute()
    }
}

node.mindMap.addListener(new MyNodeChangeListener())
node.details = "Activated: ${new Date()}"
```

So, it seems that Node Listeners are not an option.

So, just to be clear: my suggestion in my comment above is that you call firefox directly from the FP script, without sending the link to an external notepad. The activation of this script on loop can be done the way you are doing right now (AH), or you can try the multithread that I mentioned on #544. I don't have experience with this multithread solution, so I don't know how it affects the performance.

**Comment by JJSlote** [comment]
Greetings @euu2021 -- 
Many thanks for the thought and research you've put into tightening up this script pairing. Some of your findings have been built into the final (for now) version posted below, which runs with great snap and efficiency.

Yes, an all-Groovy script would ideally be triggered on a listener. I'm pretty sure, per its name, that the node change listener requires an actual update of the node to activate the script. Thus I was looking for a node selection listener; there's bound to be one in Freeplane's source, but I couldn't identify it in the API. 

On your side note, I've used Quicklook in other projects; it's a helpful throwback to the old Quickview that was included in Win95. But my HTML docs constitute an internal wiki; they link to one another and out to the web, and share a single stylesheet, so they're best presented in a full-fledged browser, one which fortunately offers an extension for Markdown viewing.

The syntax for program launch with parameters was extremely helpful!! Thanks to you and Volker.

After ditching the timer, I tried another approach that did not work out: using AHK to intercept the keys that were typically used to select a new node: the arrow keys and the left mouse click. With Freeplane active, these keys would do their own jobs and then trigger the Groovy script hotkey and the file read. But on my slow PC the timing to recognize a selected node varies quite a bit; that launch scheme proved unreliable even with (too many) sleep inserts.

The successful approach to the problem turned out to be some old-fashioned AutoHotkey screen-scraping. Freeplane's status line displays the link when it's recognized. AHK can't actually read the displayed characters absent OCR, but it can search for images quickly, particularly when limited to small sections of the screen. So my AHK script searches, three times per second, for the image of `file:///` that'll be shown when a link I want to auto-navigate is established, and that triggers the hotkey and the Groovy script activation. 

Of course the link stays visible in the status line for a stretch, and without additional filtering the hotkey will launch the Groovy script repeatedly. The AHK script can't actually read the node ID from the status line, but needs to determine whether it's fresh or stale. So it runs three more image searches in succession, this time for the glyphs for the numerals `0`, `3`, and `7` as displayed and positioned in the status line. When ImageSearch finds these glyphs, the script captures the X-coordinate and checks for changes, and roughly 999 times out of a thousand, unchanged coordinates will mean we're on the same stale node with no need to launch the Groovy script. 

Having determined the node is newly selected, I no longer need to test the node link to see if differs from the previous parm to Firefox. That makes it simpler to call Firefox directly from the Groovy script, since I won't need a global variable to capture the previous link spec for comparison. And of course calling the browser from the Groovy side means the script no longer needs to broadcast the link via the ram drive, nor the some 500 ms of wait time this took to be read reliably. In fact the script now requires no sleeping at all.

Experience has shown that iterative ImageSearches in AHK are cheap; needless calls to the Groovy script and to Firefox are dear. The price for using PNG files particular to one's Freeplane setup is, of course, the loss of generality, but that's probably par for scripting. Anyway, these are the scripts as they stand:

```
// @ExecutionModes({ON_SINGLE_NODE})
"\"C:/Program Files/Mozilla Firefox/firefox.exe\" ${node.link.text}".execute()
```

```
LaunchMMLinks:
FPZeroPrior := 0 
FPSevenPrior := 0
FPThreePrior := 0
SetTimer, MMLinks, 350
Return

MMLinks:
coordmode, pixel, Screen
ImageSearch, DocRefX, DocRefY, 64, 1019, 364, 1040, h:\FreeplaneStatusFile.png
foundimg := !errorlevel
if foundimg {
ImageSearch, FPZero, DocRefY, 64, 1019, 150, 1040, h:\FreeplaneStatusZero.png
ImageSearch, FPSeven, DocRefY, 64, 1019, 150, 1040, h:\FreeplaneStatusSeven.png
ImageSearch, FPThree, DocRefY, 64, 1019, 150, 1040, h:\FreeplaneStatusThree.png
if ((FPZero<>FPZeroPrior) or (FPSeven<>FPSevenPrior) or (FPThree<>FPThreePrior)) {
FPZeroPrior := FPZero 
FPSevenPrior := FPSeven 
FPThreePrior := FPThree
winactivate, - Freeplane - 
send, ^{F11}
}
}
Return
```
Thanks again euu for your research and constructive suggestions!!

Cheers -- Jerome


    

        

.

**Reply by euu2021** [reply]
Nice. Interesting approach. It will be useful in future scripts. A request that sometimes appears in the forum, and probably will be of my interest in the future, is handwriting. So, someday, I will try to create a quick integration with an external drawing software, like I drafted [here](https://github.com/freeplane/freeplane/discussions/593#discussioncomment-3491314).
